src/Stream.sol

```diff
@@ -237,38 +240,46 @@ contract Stream is IStream, Clone {
     function cancel() external onlyPayerOrRecipient {
         address payer_ = payer();
         address recipient_ = recipient();
-        IERC20 token_ = token();
 
         uint256 recipientBalance = balanceOf(recipient_);
+        uint256 payerBalance = balanceOf(payer_);
 
         // This zeroing is important because without it, it's possible for recipient to obtain additional funds
         // from this contract if anyone (e.g. payer) sends it tokens after cancellation.
         // Thanks to this state update, `balanceOf(recipient_)` will only return zero in future calls.
         remainingBalance = 0;
 
-        if (recipientBalance > 0) token_.safeTransfer(recipient_, recipientBalance);
-
-        // Using the stream's token balance rather than any other calculated field because it gracefully
-        // supports cancelling the stream even if payer hasn't fully funded it.
-        uint256 payerBalance = tokenBalance();
-        if (payerBalance > 0) {
-            token_.safeTransfer(payer_, payerBalance);
-        }
+        if (recipientBalance > 0) recipientTokensAfterCancel += recipientBalance;
 
         emit StreamCancelled(msg.sender, payer_, recipient_, payerBalance, recipientBalance);
     }
 
+    function withdrawRecipientTokensAfterCancel(uint256 amount) external onlyPayerOrRecipient {
+        recipientTokensAfterCancel -= amount;
```

Comment:
here we need to make sure there's enough balance before we allow the transfer :)
I'm simplifying #6 to look more like your branch, would like to continue with some extra changes there 
e.g. changing `balanceOf` such that payer's balance focuses on the stream's actual token balance rather than `remainingBalance`